### PR TITLE
Fix, and test, TCP DNS capture

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/30251

This PR has two parts:
1. Adds test for DNS in various modes, and support in echo for a dns client
2. Fixes iptables rules to properly capture TCP dns traffic.

### CODE REVIEW

File: pkg/test/echo/server/forwarder/dns.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	for _, w := range want {
+		if w == got {
+			return nil
+		}
+	}
+	return fmt.Errorf("got value %q, wanted one of %v", got, want)
+}
+
+func parseRequest(inputURL string) (dnsRequest, error) {
+	resp := dnsRequest{}
```
Comment (stevenctl) [inline pkg/test/echo/server/forwarder/dns.go:48]: nit: probably meant `req`

File: tests/integration/pilot/common/routing.go [hunk 2]
```diff
@@ -894,6 +896,126 @@ type vmCase struct {
...
+	ipv4 := "1.2.3.4"
+	ipv6 := "1234:1234:1234::1234:1234:1234"
+	dummyLocalhostServer := "127.0.0.1"
+	cases := []struct {
+		name string
+		// TODO(https://github.com/istio/istio/issues/30282) support multiple vips
+		ips      string
+		protocol string
+		server   string
+		expected []string
```
Comment (stevenctl) [inline tests/integration/pilot/common/routing.go:939]: is expected ever different from ips?

File: tests/integration/pilot/common/routing.go [hunk 3]
```diff
@@ -894,6 +896,126 @@ type vmCase struct {
...
+			name:     "udp localhost server",
+			ips:      ipv4,
+			expected: []string{ipv4},
+			protocol: "udp",
+			server:   dummyLocalhostServer,
+		},
+	}
+	for _, client := range flatten(apps.VM, apps.PodA, apps.PodTproxy) {
+		for _, tt := range cases {
+			tt, client := tt, client
```
Comment (stevenctl) [inline tests/integration/pilot/common/routing.go:983]: Can we also check `a.ns.cluster.local` to cover Service? I believe echo Instance has an `Address` method to get the cluster IP

**Review by stevenctl — COMMENTED** [review]
LGTM, letting @hzxuzhonghu take a look as well since he's assigned.

**Review by hzxuzhonghu — COMMENTED** [review]
I think this is expected request path:` app->dns-proxy->local-dns->remote dns`

The first three jumps are ok, but the last one from local dns will also be captured by dns-proxy. Seems no way to distinguish.

**Comment by howardjohn** [comment]
Yes, I have not fixed that case. Just app -> proxy -> remote dns, but using
TCP. Right now it only works for udp

**Review by hzxuzhonghu — APPROVED** [review]
LGTM, for the corner case, we have to come up with another solution

File: tests/integration/pilot/common/routing.go (hunk 2 above)
Comment (howardjohn) [reply tests/integration/pilot/common/routing.go:939]: Not right now - I had some cases where there was some but they don't actually pass today

File: tests/integration/pilot/common/routing.go (hunk 3 above)
Comment (howardjohn) [reply tests/integration/pilot/common/routing.go:983]: done

File: pkg/test/echo/server/forwarder/dns.go (hunk 1 above)
Comment (howardjohn) [reply pkg/test/echo/server/forwarder/dns.go:48]: done

**Review by stevenctl — APPROVED** [review]
